### PR TITLE
xppen: init package with nixos module

### PR DESCRIPTION
**Pull request by ghost**
## Description of changes

Added XPPen PenTablet package and a `program.xppen` module for it

## Things done

I've initially made this for my friend who asked me if I could port XPPen on NixOS, I did, so now I'd like to add it to nixpkgs.
I have a few concerns however, the primary is that I'm not quite sure about the licensing of XPPen, there is practically no information I could find and the only way to contact XPPen is through mail which, if I'm completely honest, I just don't want to do.

I also don't own an XPPen tablet and don't really know much about them unlike my friend but he's not quite there yet in terms of being able to package it himself.

Do my changes relate to #213263 and the `pentablet-driver` package???

Also this is my very first time trying to add a package/module to nixpkgs so please tell me if I'm doing things wrong, I tried my best but I suppose there are things that can be improved.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/programs/xppen.nix [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  config = mkIf cfg.enable {
+    environment.systemPackages = [
+      cfg.package
+    ];
+
+    services.udev.extraRules = ''
+      # XPPen PenTablet rules
+      KERNEL=="uinput",MODE:="0666",OPTIONS+="static_node=uinput"
+      SUBSYSTEMS=="usb",ATTRS{idVendor}=="28bd",MODE:="0666"
+    '';
```
Comment (h7x4) [inline nixos/modules/programs/xppen.nix:24]: Maybe these rules could be packaged into the package itself? I think it would be useful for anyone who's not using NixOS

File: nixos/modules/programs/xppen.nix [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+with lib; let
+  cfg = config.programs.xppen;
+in
+
+{
+
+  options.programs.xppen = {
+    package = mkPackageOption pkgs "xppen" { };
+
+    enable = mkEnableOption (lib.mdDoc "XPPen PenTablet application");
```
Comment (h7x4) [inline nixos/modules/programs/xppen.nix:12]: nit: I don't think there's any hard rules here, but the `enable` option tends to come on top in most modules, as a sort of unnamed convention.

```suggestion
    enable = mkEnableOption (lib.mdDoc "XPPen PenTablet application");
    package = mkPackageOption pkgs "xppen" { };
```

File: pkgs/tools/misc/xppen/default.nix [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+  dontConfigure = true;
+  dontBuild = true;
+  dontCheck = true;
+
+  unpackPhase = ''
+        runHook preUnpack
+        tar xzvf ${src} --strip-components=1
+        runHook postUnpack
+  '';
```
Comment (h7x4) [inline pkgs/tools/misc/xppen/default.nix:30]: Is this `unpackPhase` set manually because `fetchurl` doesn't recognize the extension? In that case, do you think you might be able to get away with using `fetchzip` and the `extension` flag instead? (the hash might need to be recalculated)

```suggestion
  src = fetchzip {
    url = "https://www.xp-pen.com/download/file.html?id=1936&pid=990&ext=gz";
    hash = "sha256-33B8JjOjsv17uFae3zhGKw40oeD6fHJ1YC9wEVQO7+Q=";
    extension = ".tar.gz";
  };

  nativeBuildInputs = [ autoPatchelfHook ];

  buildInputs = [ qt5.full libusb libGL ];

  dontConfigure = true;
  dontBuild = true;
  dontCheck = true;
```

File: pkgs/tools/misc/xppen/default.nix [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+          --replace "/usr/lib/pentablet/PenTablet.sh" "PenTablet" \
+          --replace "/usr/share/icons/hicolor/256x256/apps/xppentablet.png" "xppentablet"
+  '';
+
+  meta = with lib; {
+    description = "XPPen driver";
+    homepage = "https://www.xp-pen.com/";
+    downloadPage = "https://www.xp-pen.com/download/";
+    license = licenses.unfreeRedistributable;
+    maintainers = [ ];
```
Comment (h7x4) [inline pkgs/tools/misc/xppen/default.nix:53]: Could you add yourself as maintainer?

File: pkgs/top-level/all-packages.nix [hunk 5]
```diff
@@ -14780,6 +14780,8 @@ with pkgs;
 
   xfstests = callPackage ../tools/misc/xfstests { };
 
+  xppen = callPackage ../tools/misc/xppen { };
```
Comment (h7x4) [inline pkgs/top-level/all-packages.nix:14783]: This package might need to be moved to [`pkgs/by-name`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/README.md)

File: nixos/modules/programs/xppen.nix (hunk 1 above)
Comment (h7x4) [reply nixos/modules/programs/xppen.nix:23]: It doesn't automatically activate them unfortunately, you'll still need the module for that. But other people using nixpkgs for non-nixos systems will benefit from having these sorts of files available. Exactly how they use nixpkgs, and how they install these files is up to them.

If you install it in something like `$out/lib/udev.d/rules/50-xppen.rules`, this can be rewritten as:

```suggestion
    services.udev.packages = [ cfg.package ];
```

File: pkgs/top-level/all-packages.nix (hunk 5 above)
Comment (h7x4) [reply pkgs/top-level/all-packages.nix:14783]: It's from [RFC 140](https://github.com/NixOS/rfcs/pull/140), which was merged a while back. The plan is to move all packages into the new hierarchy at some point, but currently it's moving step by step. The CI is failing right now because #275539 was merged recently.

There's also a [nixcon talk](https://www.youtube.com/watch?v=vEpjHNZYwkA) if you're interested in a slightly less formal version of the topic. It provides some interesting historical details as well.

**Comment by h7x4** [comment]
Yes, I think so, but I'm a bit unsure about how to deal with backporting release notes. I'll ask around to see if I can find some answers.

In the meanwhile, could you squash your changes into the first three commits? If you're unsure about how to do it, feel free to ask. I might be a good idea to mark the PR as draft before push, because a rebase gone wrong will ping a ton of people.

File: pkgs/by-name/xp/xppen/package.nix [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        substituteInPlace $out/share/applications/xppentablet.desktop \
+          --replace "/usr/lib/pentablet/PenTablet.sh" "PenTablet" \
+          --replace "/usr/share/icons/hicolor/256x256/apps/xppentablet.png" "xppentablet"
+  '';
+
+  meta = with lib; {
+    description = "XPPen driver";
+    homepage = "https://www.xp-pen.com/";
+    downloadPage = "https://www.xp-pen.com/download/";
+    license = licenses.unfreeRedistributable;
```
Comment (MinerSebas) [inline pkgs/by-name/xp/xppen/package.nix:48]: You mentioned being usure about the Licensing.
Looking into the the archive at `/App/usr/lib/doc/EULA` is an End User Agreement.
The relevant parts seem to these two:
```
1.License
This EULA grants you a non-exclusive, revocable, non-transferable, limited license to download, install and use the SOFTWARE after purchasing a Hanvon Ugee product from Hanvon Ugee or its subsidiaries and affiliates ,or through its authorized sales channels, you shall strictly abide by the terms of this EULA. 
2.Restrictions
Except for the limited license expressly granted in the "License" section of this EULA, you have no other rights to this SOFTWARE . More restrictions on your use of this SOFTWARE are as follows. You must not:
A. copy or use the SOFTWARE for any purpose other than as permitted under the above section  "License".
B. modify, make derivative works of, disassemble, decrypt, reverse compile or reverse engineer any part of the SOFTWARE.
C. remove, alter or obscure any proprietary notice (including any notice of copyright or trademark) of the author or its affiliates, partners of the SOFTWARE.
```
TLDR: You may `download, install and use the SOFTWARE`, but not redistribute.
-> Use `licenses.unfree` instead

Caveat: I am not a Lawyer and this isn't legal advise. :smile:

Comment (h7x4) [reply pkgs/by-name/xp/xppen/package.nix:48]: > Does that just mean that there will be no source copies available in the cache?

Both this, and the fact that you have to explicitly set an envvar or config flag to allow for installing unfree packges.

**Comment by h7x4** [comment]
Might've been after you moved it to `pkgs/by-name`. `libusb1` is the correct one.

File: nixos/modules/programs/xppen.nix [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+  };
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = [
+      cfg.package
+    ];
+
+    services.udev.packages = [ cfg.package ];
+  };
+
```
Comment (h7x4) [inline nixos/modules/programs/xppen.nix:21]: nit:

```suggestion
{
  options.programs.xppen = {
    enable = mkEnableOption (lib.mdDoc "XPPen PenTablet application");
    package = mkPackageOption pkgs "xppen" { };
  };

  config = mkIf cfg.enable {
    environment.systemPackages = [
      cfg.package
    ];

    services.udev.packages = [ cfg.package ];
  };
```

File: pkgs/by-name/xp/xppen/package.nix [hunk 8]
```diff
@@ -0,0 +1,53 @@
+{ stdenv
+, lib
+, fetchzip
+, libusb1
+, autoPatchelfHook
+, qt5
+, libGL
+}:
+stdenv.mkDerivation rec {
```
Comment (h7x4) [inline pkgs/by-name/xp/xppen/package.nix:9]: Doesn't seem to need `rec`

```suggestion
stdenv.mkDerivation {
```

File: pkgs/by-name/xp/xppen/package.nix [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        rm -r $out/bin/lib
+        cp -r App/lib $out/lib
+        runHook postInstall
+  '';
+
+  postInstall = ''
+        substituteInPlace $out/share/applications/xppentablet.desktop \
+          --replace "/usr/lib/pentablet/PenTablet.sh" "PenTablet" \
+          --replace "/usr/share/icons/hicolor/256x256/apps/xppentablet.png" "xppentablet"
+  '';
```
Comment (h7x4) [inline pkgs/by-name/xp/xppen/package.nix:42]: We don't really have that many style guidelines yet (waiting for [RFC 166](https://github.com/NixOS/rfcs/pull/166)), but 2 space indents is part of the ones we have: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#syntax

```suggestion
  installPhase = ''
    runHook preInstall
    mkdir -p $out/bin $out/share
    cp -r App/usr/lib/pentablet/. $out/bin
    rm $out/bin/PenTablet.sh
    cp -r App/usr/share/. $out/share
    rm -r $out/bin/lib
    cp -r App/lib $out/lib
    runHook postInstall
  '';

  postInstall = ''
    substituteInPlace $out/share/applications/xppentablet.desktop \
      --replace "/usr/lib/pentablet/PenTablet.sh" "PenTablet" \
      --replace "/usr/share/icons/hicolor/256x256/apps/xppentablet.png" "xppentablet"
  '';
```

**Comment by misumisumi** [comment]
Based on the above review, if you want to replace it with another directory, you need to create the directory on the nixosModule side.
See below for details.
https://github.com/NixOS/nixpkgs/issues/213263#issuecomment-1913130606

**Comment by eclairevoyant** [comment]
> I looked through the docs, yeah, they say you shouldn't backport new modules and packages

Not sure where you saw this. [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#changes-acceptable-for-releases) explicitly says it is ok:

> Here are some examples of backwards-compatible changes that are okay to backport:
> - ✔️ New packages, modules and functions

**Comment by misumisumi** [comment]
Sorry for the confusion.
I have an xp-pen pen tablet and when I tried this package, it did not start due to an error related to xcb on the Qt platform.
Also, it probably won't be able to load the config even if it successfully boots. This is because the app tries to read files under `/usr/lib/pentablet`.

NixOS already has [xp-pen-driver packages](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/os-specific/linux/xp-pen-drivers/deco-01-v2/default.nix) for certain products, which solve the `/usr/lib` problem and the libQt5 dependency problem.
This package is a generic driver as mentioned in #213263.
Therefore, I think it would be better to deal with packages by renaming and modifying existing packages.

**Comment by misumisumi** [comment]
> is it just super old (2020 I guess)?

Yeah. The current package has outdated download links and file names and will not work as is.

> @misumisumi also could you please post the xcb error? Is [this](https://nixos.wiki/wiki/Qt#qt.qpa.plugin:_Could_not_find_the_Qt_platform_plugin_.22xcb.22_in_.22.22) related?

Yeah. That probrem.
```
qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: xcb.

[1]    1248247 IOT instruction (core dumped)  PenTablet
```

I have the modified code so send a PR to your repository

**Comment by misumisumi** [comment]
However, I have never posted a package, so I don't know if it's okay to organize existing packages.

**Comment by misumisumi** [comment]
> if nothing works then buildFHSEnv would be the way to go

Since config changed in GUI is written to `config.xml`, it is better to migrate if `buildFHSEnv` allows changing the contents of the file.

**Comment by misumisumi** [comment]
It is probably written to `/usr/lib`.
(The image shows `/var/lib`, but this is because a patch has been applied for nixos.)

image

Even if you look at the `install.sh` that exists in the tar.gz of xp-pen-driver, it just copies `App/usr/lib` to `/usr/lib`, so the config in it is I think it's changing.
In fact, the changed format and the `strace` format seem to be the same.

diff of config before and after changes
(config exists under /var/lib/pentablet, here /var/lib is not buildFHSEnv but /var/lib of the actual host and is tested with its own package based on xppen-unwrapped.)

```diff
29,32c29,32
<         <T_0 W="172" X="274" Y="72" H="98"/>
<         <T_1 W="50" X="314" Y="86" H="91"/>
<         <T_2 W="172" X="233" Y="72" H="98"/>
<         <T_3 W="51" X="315" Y="66" H="91"/>
---
>         <T_0 H="98" W="172" X="274" Y="72"/>
>         <T_1 H="91" W="50" X="314" Y="86"/>
>         <T_2 H="98" W="172" X="233" Y="72"/>
>         <T_3 H="91" W="51" X="315" Y="66"/>
37c37
<         <PenBtn1 id="0" Actid="207"/>
---
>         <PenBtn1 Actid="207" id="0"/>
46,51c46,51
<         <K1 Show="1" id="0" Motid="1" Actid="1"/>
<         <K2 Show="1" id="0" Motid="2" Actid="2"/>
<         <K3 Show="1" id="0" Motid="3" Actid="3"/>
<         <K4 Show="1" id="0" Motid="4" Actid="6"/>
<         <K5 Show="1" id="0" Motid="5" Actid="101"/>
<         <K6 Show="1" id="0" Motid="6" Actid="105"/>
---
>         <K1 Show="1" Actid="1" Motid="1" id="0"/>
>         <K2 Show="1" Actid="2" Motid="2" id="0"/>
>         <K3 Show="1" Actid="3" Motid="3" id="0"/>
>         <K4 Show="1" Actid="6" Motid="4" id="0"/>
>         <K5 Show="1" Actid="101" Motid="5" id="0"/>
>         <K6 Show="1" Actid="105" Motid="6" id="0"/>
115,118c115,118
<         <T_0 W="163" X="272" Y="69" H="102"/>
<         <T_1 W="53" X="313" Y="87" H="86"/>
<         <T_2 W="163" X="241" Y="71" H="102"/>
<         <T_3 W="53" X="313" Y="69" H="86"/>
---
>         <T_0 H="102" W="163" X="272" Y="69"/>
>         <T_1 H="86" W="53" X="313" Y="87"/>
>         <T_2 H="102" W="163" X="241" Y="71"/>
>         <T_3 H="86" W="53" X="313" Y="69"/>
123,124c123,124
<         <PenBtn1 id="0" Actid="207"/>
<         <PenBtn2 id="0" Actid="209"/>
---
>         <PenBtn1 Actid="207" id="0"/>
>         <PenBtn2 Actid="209" id="0"/>
133,138c133,138
<         <K1 Show="1" id="0" Motid="7" Actid="1"/>
<         <K2 Show="1" id="0" Motid="8" Actid="2"/>
<         <K3 Show="1" id="0" Motid="1" Actid="6"/>
<         <K4 Show="1" id="0" Motid="5" Actid="21"/>
<         <K5 Show="1" id="0" Motid="3" Actid="22"/>
<         <K6 Show="1" id="0" Motid="4" Actid="101"/>
---
>         <K1 Show="1" Actid="1" Motid="7" id="0"/>
>         <K2 Show="1" Actid="2" Motid="8" id="0"/>
>         <K3 Show="1" Actid="6" Motid="1" id="0"/>
>         <K4 Show="1" Actid="21" Motid="5" id="0"/>
>         <K5 Show="1" Actid="22" Motid="3" id="0"/>
>         <K6 Show="1" Actid="101" Motid="4" id="0"/
```

**Comment by misumisumi** [comment]
Where can I find `PenTabletrc`?
I tried searching in `strace` but couldn't find it.
Where can I see `PenTabletrc` in strace?
I tried searching in `strace` but couldn't find it.
Also, no such file or directory exists under XDG_CONFIG_PATH

**Comment by misumisumi** [comment]
Hmm.
I am running the same binary (`xppen`), but it does not appear in my strace.
Is it because my environment is not KDE?

**Comment by misumisumi** [comment]
Yeah. Exist.
image

**Comment by misumisumi** [comment]
It will start up and temporary settings can be made.
However, the settings are not saved.
image

**Comment by misumisumi** [comment]
OK.
I thought it would be best if we could write to `PenTabletrc` separately from `config.xml`, so I installed it on the VM's ubuntu and checked, but it didn't appear in strace, so I was just curious.
I'll check it on the KDE system tomorrow.

**Comment by misumisumi** [comment]
I checked the driver installed on the KDE Neon VM, but I still couldn't find the `PenTabletrc` string. (Need to keep in mind that the tablet is not connected)

I investigated the behavior in various communities, and it seems that at least in older versions, the specification was to write directly to files under `/usr/`. https://www.reddit.com/r/XPpen/comments/12xgoq0/linux_driversutility_the_most_insecure_piece_of/

Anyway, as you say, `buildFHSEnv` is a pain, so I think it's okay to push it here.

**Comment by misumisumi** [comment]
It will fail if uinput is not present, so please add the following to `nixosModule`

```nix
     hardware.uinput.enable = true;
````

File: nixos/modules/programs/xppen.nix [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+  config = mkIf cfg.enable {
+    hardware.uinput.enable = true;
+
+    environment.systemPackages = [
+      cfg.package
+    ];
+
+    services.udev.packages = [ cfg.package ];
+
+    system.activationScripts.xppen.text = with pkgs; ''
```
Comment (MinerSebas) [inline nixos/modules/programs/xppen.nix:22]: Does this need to be a `activationScripts`?
I think, a systemd oneshot service, that would only need to be restarted if `xppen` changed would be better, than a activationScripts that needs to always run on activation.

File: nixos/modules/programs/xppen.nix [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+      install -m 755 -d "/var/lib/pentablet/conf/xppen"
+
+      readarray -d "" files < <(find ${xppen}/usr/lib/pentablet/conf -type f -print0)
+
+      for file in "''${files[@]}"; do
+        file_new="/var''${file#${xppen + "/usr"}}"
```
Comment (MinerSebas) [inline nixos/modules/programs/xppen.nix:28]:
```suggestion
      readarray -d "" files < <(find ${cfg.package}/usr/lib/pentablet/conf -type f -print0)

      for file in "''${files[@]}"; do
        file_new="/var''${file#${cfg.package + "/usr"}}"
```

**Comment by MinerSebas** [comment]
The missing piece should just be adding `restartTriggers = [ cfg.package ]` to the service.

File: nixos/modules/programs/xppen.nix [hunk 12]
```diff
@@ -0,0 +1,44 @@
+{ config, lib, pkgs, ... }:
+
+with lib; let
```
Comment (nu-nu-ko) [inline nixos/modules/programs/xppen.nix:3]:
```suggestion
let
```
to address #208242 for this module, uses of `with` such as this should be avoided.
consider `inherits` where you find yourself using `lib.foo` excessively.

File: nixos/modules/programs/xppen.nix [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+  cfg = config.programs.xppen;
+in
+
+{
+  options.programs.xppen = {
+    enable = mkEnableOption (lib.mdDoc "XPPen PenTablet application");
```
Comment (nu-nu-ko) [inline nixos/modules/programs/xppen.nix:9]:
```suggestion
    enable = mkEnableOption "XPPen PenTablet application";
```
`lib.mdDoc` is now just an alias and can be safely removed everywhere.
see https://github.com/NixOS/nixpkgs/commit/d36f950d405369abca5144b14a201bc25c4a080a and https://github.com/NixOS/nixpkgs/pull/237557

File: nixos/modules/programs/xppen.nix [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+    package = mkPackageOption pkgs "xppen" { };
+  };
+
+  config = mkIf cfg.enable {
+    hardware.uinput.enable = true;
+
+    environment.systemPackages = [
+      cfg.package
+    ];
```
Comment (nu-nu-ko) [inline nixos/modules/programs/xppen.nix:18]:
```suggestion
    environment.systemPackages = [ cfg.package ];
```

**Review by fsnkty — CHANGES_REQUESTED** [review]

File: nixos/modules/programs/xppen.nix [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+
+    services.udev.packages = [ cfg.package ];
+
+    systemd.services.xppen-create-config-dir = {
```
Comment (eclairevoyant) [inline nixos/modules/programs/xppen.nix:22]: This service seems like something that might be better achieved with systemd-tmpfiles

File: pkgs/by-name/xp/xppen/package.nix [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+    rm $out/bin/PenTablet.sh
+    rm -r $out/bin/conf
+    rm -r $out/bin/doc
+    rm -r $out/bin/lib
+    rm -r $out/bin/platforms
+    runHook postInstall
+  '';
+
+  postInstall = ''
+    sed -i 's#/usr/lib/pentablet#/var/lib/pentablet#g' $out/bin/PenTablet
```
Comment (eclairevoyant) [inline pkgs/by-name/xp/xppen/package.nix:48]: prefer the substitute* functions over `sed`

File: pkgs/by-name/xp/xppen/package.nix [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+
+    substituteInPlace $out/share/applications/xppentablet.desktop \
+      --replace "/usr/lib/pentablet/PenTablet.sh" "PenTablet" \
```
Comment (eclairevoyant) [inline pkgs/by-name/xp/xppen/package.nix:51]:
```suggestion
      --replace-fail "/usr/lib/pentablet/PenTablet.sh" "PenTablet" \
```

File: pkgs/by-name/xp/xppen/package.nix [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+      --replace "/usr/share/icons/hicolor/256x256/apps/xppentablet.png" "xppentablet"
```
Comment (eclairevoyant) [inline pkgs/by-name/xp/xppen/package.nix:52]:
```suggestion
      --replace-fail "/usr/share/icons/hicolor/256x256/apps/xppentablet.png" "xppentablet"
```

**Review by eclairevoyant — DISMISSED** [review]
Thank you for your work so far! Some feedback below.

File: nixos/modules/programs/xppen.nix (hunk 15 above)
Comment (eclairevoyant) [reply nixos/modules/programs/xppen.nix:20]: tmpfiles are not necessarily temporary, despite the name.

File: pkgs/by-name/xp/xppen/package.nix (hunk 16 above)
Comment (eclairevoyant) [reply pkgs/by-name/xp/xppen/package.nix:48]: oh, it's an ELF, fair enough.

**Comment by damiankorcz** [comment]
Hey,
I have an XP-Pen Deco LW. This model works wired and wireless so I can test both scenarios. It does get detected and moves the cursor when using KDE 6 wired or wireless. I would like to help get this updated and merged. I've tried to build it and it appears to build fine although when running the PenTablet executable I get the following:
```bash
[damian@nixos-desktop:~/Git/nixpkgs/result/bin]$ ./PenTablet
QFont::fromString: Invalid description 'Noto Sans,10,-1,5,400,0,0,0,0,0,0,0,0,0,0,1'
QLocale::English
QMetaObject::connectSlotsByName: No matching signal for on_horizontalSlider_tablet_light_valueChanged(int)
QObject::connect: Incompatible sender/receiver arguments
        QDesktopWidget::primaryScreenChanged() --> MainWindow::onDeskScreenChanged(int)
QLayout: Attempting to add QLayout "" to MainWindow "MainWindow", which already has a layout
Function  EnumDisplay nScreenCount: 4
Device.AllScreen "(0,0,4640,2880)"
Device.Screen 1 : "1(1440,1440,2560,1440)"
Device.Screen 2 : "2(0,320,1440,2560)"
Device.Screen 3 : "3(1440,0,2560,1440)"
Device.Screen 4 : "4(4000,2400,640,480)"
-------type: 1
hid_open_path Failed!
Cannot mix incompatible Qt library (5.15.14) with this library (5.15.12)
Aborted (core dumped)
```
Let me know what you want me to try further.

**Comment by gepbird** [comment]
A few months ago drivers with major version 4 got released. However some tablets [are not compatible with v4](https://aur.archlinux.org/packages/xppenlinux-v3#comment-977654), so we should definitely keep the current v3 driver.

Based on the previous v3 deco and g430 drivers, I packaged a working v4 driver in https://github.com/NixOS/nixpkgs/pull/347984 (tested by @HangedFool, I don't have a hardware for it), however it is not as polished as this PR but should be a decent starting point.

It would be nice to add the v4 package in this PR, but I totally understand if you don't have the capacity for it, it can be done later. At least we should make the v3 package in this PR future proof by renaming it to `xppen_3` to avoid making aliases in the future that might confuse/annoy users.

**Comment by gepbird** [comment]
> ```shell
> Cannot mix incompatible Qt library (5.15.14) with this library (5.15.12)
> ```

@damiankorcz This is most likely not an issue with the the xppen package in this PR, but rather with Qt packages on NixOS in general.

Personally I don't have this issue and the GUI opens, but this is what I found after doing some research:
- according to https://github.com/NixOS/nixpkgs/issues/37864#issuecomment-516164854, the issue should be fixed by using `wrapQtAppsHook`, but it's already done with xppen so it's not the case
- (temporarily) uninstalling Qt apps and removing them from the nix store worked for some people, see https://discourse.nixos.org/t/migrating-from-plasma5-to-plasma6/40514/3
- people strace'd the problematic Qt apps to find out why is it trying to use different Qt versions https://github.com/NixOS/nixpkgs/issues/30551#issuecomment-337903477

I haven't found a clear answer, but if you really want to test this PR you could try temporarily uninstalling your qt apps (also delete them with nix-collect-garbage) or updating your system to unstable (I assume you're on nixos-24.05 from the 5.15.12 Qt version). Or at this point setting up a nixos-unstable vm seems to be easier.

File: pkgs/by-name/xp/xppen/package.nix [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+, lib
+, fetchzip
+, autoPatchelfHook
+
+, libusb1
+, qt5
+, libGL
+}:
+stdenv.mkDerivation {
+  pname = "xppen";
```
Comment (gepbird) [inline pkgs/by-name/xp/xppen/package.nix:11]: See https://github.com/NixOS/nixpkgs/pull/285660#issuecomment-2411961311 for explanation

```suggestion
  pname = "xppen_3";
```

File: nixos/modules/programs/xppen.nix [hunk 20]
```diff
@@ -0,0 +1,42 @@
+{ config, lib, pkgs, ... }:
+
+let
+  cfg = config.programs.xppen;
+in
+
+{
+  options.programs.xppen = {
+    enable = lib.mkEnableOption "XPPen PenTablet application";
+    package = lib.mkPackageOption pkgs "xppen" { };
```
Comment (gepbird) [inline nixos/modules/programs/xppen.nix:10]:
```suggestion
    package = lib.mkPackageOption pkgs "xppen_3" { };
```

**Review by gepbird — CHANGES_REQUESTED** [review]
Hi, I don't want to be intrusive, but if you don't have time to work on this I'd be happy to take it over if it's fine with you (of course I'd credit you for all the work you've done over the past months :) ).

Is there anything else blocking this PR?
